vochain/account.go

```diff
@@ -692,6 +698,97 @@ func (v *State) SetAccountDelegate(accountAddr, delegateAddr common.Address, txT
 	}
 }
 
+// CollectFaucetTxCheck checks if a CollectFaucetTx and its data are valid
+func CollectFaucetTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) (*common.Address, error) {
+	if vtx == nil {
+		return nil, ErrNilTx
+	}
+	tx := vtx.GetCollectFaucet()
+	// check signature available
+	if signature == nil || tx == nil || txBytes == nil {
+		return nil, fmt.Errorf("missing signature and/or transaction")
+	}
+	faucetPkg := tx.GetFaucetPackage()
+	// check faucet pkg content
+	if faucetPkg == nil {
+		return nil, fmt.Errorf("nil faucet package")
+	}
+	if faucetPkg.Signature == nil {
+		return nil, fmt.Errorf("invalid faucet package signature")
+	}
+	if faucetPkg.Payload == nil {
+		return nil, fmt.Errorf("invalid faucet package payload")
+	}
+	if faucetPkg.Payload.Amount == 0 {
+		return nil, fmt.Errorf("invalid faucet package payload amount")
+	}
+	// recover txSender address from signature
+	pubKey, err := ethereum.PubKeyFromSignature(txBytes, signature)
+	if err != nil {
+		return nil, fmt.Errorf("cannot extract public key from signature: %w", err)
+	}
+	toAddr, err := ethereum.AddrFromPublicKey(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("cannot extract address from public key: %w", err)
+	}
+	// extract the account that generated the payload
+	faucetPackageBytes, err := proto.Marshal(faucetPkg.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("cannot extract faucet package payload: %v", err)
```

Comment:
why %v and not %w?

(maybe it's correct, there's also another %v a couple of returns later, in line 750)

Reply:
is correct, but we wrap all errors at this level, changed all of them